Adding support for CuDNN-based LSTM with projections

Kipok opened the pull request:
Fixes #46213

I didn't yet update the documentation, will add those change soon. A few other things that I didn't do, but want to clarify if I maybe should.

1. I didn't expose projections in c++ API: torch/csrc/api/src/nn/modules/rnn.cpp. Let me know if this is desirable and I will add those changes.
2. I didn't expose projections in "lstm_cell" function and "_thnn_differentiable_lstm_cell_backward" functions from aten/src/ATen/native/RNN.cpp. As far as I understand, they are not needed for nn.LSTM CPU execution. For lstm_cell, projections don't bring any real benefit, since if cell is used separately, it can be easily added in Python. For "_thnn_differentiable_lstm_cell_backward", I'm actually not sure where exactly that function is used, so I also disabled projections there for now. Please let me know if I should change that.
3. I added check that projections are not supported for quantized LSTMs to quantized_lstm_<data/input> functions. But I didn't add any checks to LSTMCell code. It seems that since I disabled projections in "lstm_cell" function, they should also not be available for quantized models through any other API than quantized_lstm_<data/input>. Please let me know if I'm not correct and I will add checks to other places.
4. Projections are not supported for CuDNN versions < 7.1.2. Should I add the check for CuDNN version and disable projections in that case? If so, what will be the best way to do that?
5. Currently I added projection weight as the last weight, so the layout is "w_ih, w_hh, b_ih, b_hh, w_hr". This breaks the assumption that biases come after weights and thus I had to add additional if-s in various places. Alternative way would be to have "w_ih, w_hh, w_hr, b_ih, b_hh" layout, in which case the assumption will be true. But in that case I will need to split the loop in get_parameters function from aten/src/ATen/native/cudnn/RNN.cpp. And in some cases, I will still need to add an "undefined" tensor in the 3rd position, because we get all 5 weights from CuDNN most of the time. So I'm not sure which way is better. Let me know if you think I should change to the weights-then-biases layout.

Commit:
Expose proj_dim parameter

Commit:
Fix int -> int_64t issue

Commit:
Start exposing proj_size throughout the code base

Commit:
Exposed through get_weight_buf and flatten_weights

Commit:
Expose proj_size through _cudnn_rnn

Commit:
Update get_parameters to work with projections

Commit:
Remove redundant proj_size, fix get_expected_data_ptrs

Commit:
Fix try_get_weight_buf

Commit:
Fix _cudnn_rnn function

Commit:
Fix _cudnn_rnn_backward functions

Commit:
Add correct hx creation on python side

Commit:
Fix incorrect projection layers initialization

Commit:
Correct weight initialization on python side

Commit:
Fix output size issue

Commit:
Fix multi-layer projections issue

Commit:
Expose proj_size in setstate and extra_repr

Commit:
Fix AutocastRNN to accept models with projections

Commit:
Merge branch 'master' into cudnn_projections

Commit:
Add test TODOs, add check for non-cudnn code

Commit:
Fix incorrect hidden states init for LSTM

Commit:
Fix error for RNN/GRU of accessing undefined cx

Commit:
Fix no-bias projections lstm for fp32

Commit:
Fix no-bias projection fp16 case

Commit:
Add check for rnn/gru, add initial unit tests

Commit:
Add proj_size to test_variable_sequence

Commit:
Add projections to rnn_weight_norm test

Commit:
Add projections to cudnn_weight_format test

Commit:
Add projections to cudnn_weight_tying test

Commit:
Add projections to rnn_args_check test

Commit:
Add projections to rnn_check_device test

Commit:
Expose cudnn with projections on CPU

Commit:
Add cpu_vs_gpu projections test

Commit:
Add weight norm test

Commit:
Remove TODOs

Commit:
Code clean up

Commit:
Revert empty changes

Commit:
Revert miopen style change

Commit:
Disable projections for quantized LSTMs

Commit:
Merge branch 'master' into cudnn_projections

glaringlee reviewed (commented):
@Kipok 
Thanks for your contribution. 
There are several things here:
1. You directly changed the function signature in native_functions.yaml, this is not backward compatible IMO (BC Breaking), I think we probably need to add an overload version, for eg. `_cudnn_rnn.project_size(...)` or need to confirm this is not BC Breaking. @zou3519 

2. C++ API expose, yes, it would be great if you can add it to C++ API as well. But not necessary in this PR, we can do it in next PR if you would like to.

3. For the rest of your questions in the PR description. @zou3519 @albanD.
 
4. Your code currently broke several CI tests as you can see in CI test section. Most of them are caused by not updating the related test scripts, you can find them in the error log, please fix them (plus lint error) after above questions are cleared. 

@zou3519 @albanD can you guys help answering questions in the PR description related to this RNN changes? And appreciated if you guys can help reviewing this code a bit. thanks.

mcarilli commented:
Re 5., I don't think you have a choice.  I think cudnn itself expects/requires a particular layout sequence for flattened weights.

Kipok commented:
Thanks for the feedback @glaringlee. Concerning changing function signatures, I can probably remove proj_size argument from there, so that there is no change. It can be inferred from hx/cx Tensors passed to those functions. If I do that, it will not be possible to check that hx/cx are of correct size, but it's probably ok given that these checks are also done in the higher level API. Let me know if you think it's ok to do that and I will then remove proj_size to keep backward compatibility.

Commit:
Fix linting errors

Commit:
Remove .cuda() call from proj hidden_state test

Commit:
Add projections documentation to nn.LSTM

Commit:
Merge branch 'master' into cudnn_projections

Kipok commented:
Fixed some of the failing tests and added documentation. I only modified nn.LSTM docs, please let me know if any other place in the documentation needs to be modified as well. Attached are the screenshots of the rendered documentation changes.
![screen1](https://user-images.githubusercontent.com/2354422/98860551-b803bd80-2418-11eb-9e30-a6db09802074.png)
![screen2](https://user-images.githubusercontent.com/2354422/98860761-00bb7680-2419-11eb-9d14-323c2ed3de58.png)

Commit:
Remove cuda placement from proj initial_hidden_state test

Kipok commented:
@zou3519, @ngimel, any feedback on this PR? There are currently 3 failing tests, one of which will be fixed if I remove "proj_size" from function definitions and infer it from the states shape (see above on my comments for that). Please let me know if you think it's ok to do that and I will change the code. The other 2 tests are related to onnx, and I'm not really sure how to fix them. Any feedback on that will be very helpful.

zou3519 commented:
I've never been able to run the onnx tests locally. cc @houseroad do you know if there's a guide somewhere to running the onnx tests locally?

ngimel commented on aten/src/ATen/native/cudnn/RNN.cpp:
```diff
@@ -576,14 +681,26 @@ namespace {
   }
 
   std::vector<int64_t> _hidden_size(const RNNDescriptorParams& rnn, const TensorDescriptorListParams& tensors) {
+    if (rnn.proj_size != 0 && rnn.proj_size != rnn.hidden_size) {
```
you are checking rnn.proj_size != rnn.hidden_size in multiple places, but in rnn.py you explicitly assert that proj_size is less than hidden_size, so it can never be true?

Kipok replied:
So, there are actually 2 ways how "no projection" LSTM can currently be defined. First, if proj_size is explicitly set to 0. This way assumes that proj_size is passed directly as an argument to a function and not inferred from cell dimensions. But in fact the main point of entry to all these functions is through "_cudnn_impl" function, which infers "proj_size" from hidden dimensions. In that case was easier for me to just say that "proj_size = hx.size(2); hidden_size = cs.size(2)" and if they are not equal, we will have LSTM with projections, if they are equal, it's usual LSTM. I agree that it may be misleading (which you also point out in the later comment). Let me try to unify this convention and always explicitly assign proj_size to 0 when it is inferred and hidden dimensions are equal

ngimel commented on aten/src/ATen/native/cudnn/RNN.cpp:
```diff
@@ -700,21 +822,27 @@ namespace cudnn_rnn {
-          orig_param.set_(new_param.view_as(orig_param));
+        // There is a special case for LSTM with projections and no bias,
+        // where weight copy is done in 0->0, 1->1, 2->4 layout
+        if (weight[i].size() == 3) {
```
wouldn't weight[i].size() be 3 also for GRU with no bias?

Kipok replied:
Added check for params.size as well to distinguish these cases

ngimel commented on aten/src/ATen/native/cudnn/RNN.cpp:
```diff
@@ -1321,22 +1469,30 @@ std::pair<Tensor, hidden_type> _cudnn_impl(
+  if (cx.defined()) {
+    hidden_size = cx.size(2);
+  }
+  int64_t proj_size = hx.size(2);
```
This is misleading for a case when there's no projection (instead of 0 it will be computed as `hx.size(2)`

ngimel reviewed (commented):
Generally looks good, I did not look at the tests yet. 
In addition to comments, there's also a function 
```
_thnn_differentiable_lstm_cell_backward
```
that makes sure that double backward can be run on lstm cells (it's tested somewhere in test_autograd I believe). With projection, it is no longer correct. You should either error out there, saying that double backward is not supported with projection, or fix the formulas. 
Edit: actually, additional matrix multiply is after _thnn_fused_lstm_cell, so autograd is probably fine, it should be able to track additional matrix multiply and produce gradients for it.

Commit:
Merge branch 'master' into cudnn_projections

Kipok commented on aten/src/ATen/native/cudnn/RNN.cpp:
```diff
@@ -378,6 +379,62 @@ namespace {
+  ) {
+    void* matrix_pointer = nullptr;
+    // assuming it's LSTM which has 8 layers
+    int64_t linear_id = 8;
```
I agree that calling this "layers" is misleading. But I was just following the convention used in other parts of the code, e.g. "_num_linear_layers" function defined in this file or "num_linear_layers" variable used in "get_parameters" function. Let me explain this in the comment here, so that it's clearer. For 8 as projection weight, I actually took it from TensorFlow implementation for CuDNN LSTM. Not sure if this is the best way to do it, but it seems to be working.

Commit:
Address PR comments

Commit:
Expose projections in c++ API

Commit:
Add c++ integration tests for projections

Commit:
Add check output size test

Commit:
Add more projection tests to c++ api

Kipok commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -68,6 +68,11 @@ using CellParamsSerializationType = std::tuple<
   virtual Tensor matmul_hh(const Tensor& h) const = 0;
+  // by default doing nothing. CellParams will override this
+  // to define correct behavior for LSTMs with projections
+  virtual Tensor matmul_hr(const Tensor& h) const {
```
The reason is that it's easier to define it here, so that this default behavior can be directly re-used in all function that don't support projections (e.g. quantized cells). I added explanation to this in the code comments as well

Kipok commented:
Thanks @ngimel for the review! I have addressed your comments, please have another look at the code. I have also exposed projections implementation to c++ API and added corresponding tests.

For some reason I'm getting flake8 errors for the files that I didn't modify. I did merge with latest master, though. Could it be that those files are raising errors on master as well?

I also realized that it might be complicated for me to remove proj_size argument from "_cudnn_rnn_flatten_weight" function. So, I will either need to keep it as is (which means it will technically not be backward compatible) or create a new function just for the projection case. Given that it starts with underscore, is it maybe ok to break backward compatibility here? If not, please let me know and I will create a new function.

Kipok commented:
Also for the onnx tests -- I can easily fix them by blindly changing the failing asserts to have new values. But I don't have enough knowledge about onnx to understand what is going on there -- honestly no idea why those tests are impacted in any way, given that they do not use projections. On the other hand, is some change required to support LSTMs with projections on the onnx side? Or at least have some way to error-out on the conversion, if it's not going to work out-of-the-box?

Kipok commented:
@zou3519, @ngimel, can you please help to review the new code version? Also any comments about test cases failures will be very helpful. Thanks!

zou3519 commented on torch/nn/modules/rnn.py:
```diff
@@ -34,10 +34,11 @@ class RNNBase(Module):
     batch_first: bool
     dropout: float
     bidirectional: bool
+    proj_size: bool
```
This should be int, right?

Kipok replied:
Right, thanks for catching this!

zou3519 commented:
The flake8 failures seem unrelated, maybe try rebasing this PR on top of master (or viable/strict).

The rest of the type failures are from type hints. PyTorch runs the mypy type checker on a bunch of code to make sure the types are reasonable. You can reproduce the test failures locally either via `python test/test_type_hints.py -v` or by just running `mypy` at the top level but I've also added some inline comments on them.

zou3519 commented on torch/nn/modules/rnn.py:
```diff
@@ -75,20 +81,33 @@ def __init__(self, mode: str, input_size: int, hidden_size: int,
                 # bias vector is needed in standard definition.
                 b_hh = Parameter(torch.Tensor(gate_size))
-                layer_params = (w_ih, w_hh, b_ih, b_hh)
+                if self.proj_size == 0:
+                    if bias:
+                        layer_params = (w_ih, w_hh, b_ih, b_hh)
+                    else:
+                        layer_params = (w_ih, w_hh)
```
I think mypy is just complaining that layer_params can be a bunch of different things.

Try adding the following before the if statement:
```
layer_params: Tuple[Tensor, ...] = ()
```

zou3519 commented on torch/nn/modules/rnn.py:
```diff
@@ -531,15 +588,24 @@ class LSTM(RNNBase):
 
+    def get_expected_cell_size(self, input: Tensor, batch_sizes: Optional[Tensor]) -> Tuple[int, int, int]:
+        if batch_sizes is not None:
+            mini_batch = batch_sizes[0]
+            mini_batch = int(mini_batch)
```
> torch/nn/modules/rnn.py:594: error: Incompatible types in assignment (expression has type "int", variable has type "Tensor")  [assignment]
torch/nn/modules/rnn.py:596: error: Incompatible types in assignment (expression has type "int", variable has type "Tensor")  [assignment]

The error message is complaining that mini_batch had type Tensor, but then we reassigned it a value of type int.
```
mini_batch = int(batch_sizes[0])
```
should fix that

Commit:
Merge branch 'master' into cudnn_projections

Commit:
Add correct type hints

Kipok commented:
Thanks @zou3519! I addressed your comments. Still need some guidance on what to do with backward compatibility and onnx tests.

ngimel commented:
Backward compatibility allows adding arguments with default values to the end of the signature, so there are 2 options
1) move proj_size to the end, specify default value of 0
2) do a bc breaking change, add the operators to the allow-list. 
Pros of 1) is no bc-breaking change, but the signatures are obviously going to become more awkward. 
2) is not out of question, because as far as I see breakage is only for internal _cudnn function, so the consequences are not awful. Thoughts?

ngimel commented:
Thanks for addressing my comments, looks good! I've also looked over the tests, and they are pretty thorough. As for onnx tests, are you able to run them locally? I did not look over all failures, but some of them looked like they should be fixed in the tests themselves (e.g. asserting number of parameters or number of arguments). 
@houseroad @eellison do you know who can help with onnx tests?

eellison commented:
cc @BowenBao @neginraoof

Kipok commented:
Thanks @ngimel for looking at the PR one more time! 

For the onnx tests, I can run them locally, but I'm just afraid of simply changing the numbers in asserts, since I feel that something fundamental might need to be changed to properly support LSTM with projections. But of course, I can always just change asserts (I verified that if I blindly change the numbers, all the tests are passing).

For the backward compatibility, I personally would prefer option 2, it seems that changing internal functions shouldn't have too much impact, but we will get benefit of keeping signatures cleaner. But I don't know if those functions are used in some downstream applications, so it's easy for me to change signature or even create new functions if it's a better approach. Just need some guidance on this from you and others on PyTorch team.

BowenBao commented:
The failures are from caffe2 tests. From the look of the failure, it seems it's failing a strict check on expected number of nodes. Changing number to assert may be legit since code changes in rnn.py can affect the final traced graph.

@Kipok could you check/compare the exported graph before and after your change to see if the difference is expected?

```
Dec 04 23:37:41 self = <test_pytorch_onnx_caffe2.TestCaffe2Backend_opset7 testMethod=test_rnn_init_predict_split>
Dec 04 23:37:41 
Dec 04 23:37:41     def test_rnn_init_predict_split(self):
Dec 04 23:37:41         model = nn.LSTM(RNN_INPUT_SIZE, RNN_HIDDEN_SIZE, 3, bidirectional=True)
Dec 04 23:37:41         seq_lengths = np.random.randint(1, RNN_SEQUENCE_LENGTH + 1, size=7)
Dec 04 23:37:41         seq_lengths = list(reversed(sorted(map(int, seq_lengths))))
Dec 04 23:37:41         input = [torch.randn(l, RNN_INPUT_SIZE) for l in seq_lengths]
Dec 04 23:37:41         input = rnn_utils.pad_sequence(input)
Dec 04 23:37:41     
Dec 04 23:37:41         # Test that we are correctly splitting between init and
Dec 04 23:37:41         # predict net. When we embed parameters, there should be more
Dec 04 23:37:41         # ops in the init net.
Dec 04 23:37:41         mp = onnx.ModelProto.FromString(do_export(model, input, export_params=self.embed_params,
Dec 04 23:37:41                                                   keep_initializers_as_inputs=True,
Dec 04 23:37:41                                                   do_constant_folding=False)[0])
Dec 04 23:37:41         prepared = c2.prepare(mp, device='CPU')
Dec 04 23:37:41         if self.embed_params:
Dec 04 23:37:41             assert len(prepared.init_net.op) == 875
Dec 04 23:37:41             assert len(prepared.predict_net.op) == 130
Dec 04 23:37:41         else:
Dec 04 23:37:41 >           assert len(prepared.init_net.op) == 8
Dec 04 23:37:41 E           AssertionError: assert 12 == 8
Dec 04 23:37:41 E             +12
Dec 04 23:37:41 E             -8
```

BowenBao commented:
The test is asserting on number of ops from caffe2 net loaded from exported onnx model... 
```
init_net, predict_net = cls._onnx_model_to_caffe2_net(model, device, opset_version, False)
```
From onnx perspective the changes are expected due to change in rnn.py. I think it is fine to update the assertion number. @eellison, @houseroad for awareness on caffe2 side.

Commit:
Change number of ops in caffe2 onnx tests

Commit:
Add unimplemented call for LSTMs with projections in onnx

Commit:
Add onnx test to check projections not supported

Commit:
Merge branch 'master' into cudnn_projections

Kipok commented:
Thanks @BowenBao for helping with onnx questions! I have changed the asserts and also added error message for the onnx export as LSTMs with projections are not supported there (will require modifying onnxruntime code to enable this support).

I think now only backward compatibility question remains open. @ngimel, @zou3519, please let me know which option you prefer here. If you are ok with breaking backward compatibility in this case, please let me know how to "add the operators to the allow-list".

ngimel commented:
Let's try breaking bc, option 2. To make bc tests pass, you should add the functions you are breaking to `allow_list` in test/backward_compatibility/check_backward_compatibility.py with some future date (say, couple weeks ahead), look at the examples of functions already there.

Commit:
Add _cudnn_rnn functions to allow_list for bc

ngimel commented:
Importing to see what internal CI says.

Kipok commented:
@ngimel, @zou3519, both onnx and bc tests are passing now, but there are still 2 failures related to some specific builds. Any suggestions on what needs to be done to fix those? Or is it ok that they fail?

ngimel commented:
Bazel build looks unrelated, probably should go away with rebase. ROCm build should be fixed - you should modify projection tests to expect to raise RuntimeError on ROCm, and adjust tolerance for the failing fp16 test.

Commit:
Adjust tests to work on rocm

Commit:
Merge branch 'master' into cudnn_projections

Kipok commented:
@ngimel, what would be the right way to test if ROCm is used? I tried to add "TEST_WITH_ROCM" if, but it seems that in some cases it's still using cudnn? Is it possible to check it on the model level itself somehow, e.g. like I can distinguish between cpu and cuda by checking .device() property?

Commit:
Add more precise check for runtime error in tests

Commit:
Merge branch 'master' into cudnn_projections

Kipok commented:
@ngimel, @zou3519, I fixed ROCm tests, no everything passes, except for some test case that seems to be unrelated to my PR (at least it was not failing on the previous test runs). Is there anything else that I need to do to complete the merge? Should I rebase with master one more time to check if the failing test will be fixed?

ngimel commented:
asan failure is unrelated. Let me try importing, thank you!
